Fix waiting setup tasks when they are not a direct upstream

hussein-awala opened the pull request:
<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->
closes: #33561

This PR force the TI to wait all the upstream setup tasks event if they are not a direct upstream. Please check #33561 for more details.


<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

hussein-awala committed:
Fix waiting setup tasks when they are not a direct upstream

dstandish commented on airflow/ti_deps/deps/trigger_rule_dep.py:
```diff
@@ -234,14 +250,19 @@ def _iter_upstream_conditions() -> Iterator[ColumnOperators]:
 
         upstream_done = done >= upstream
+        setup_done = (success_setup + skipped_setup + failed_setup) >= upstream_setup
+        is_tear_down = task.is_teardown or trigger_rule == TR.ALL_DONE_SETUP_SUCCESS
```
i don't think you need to check both is_teardown and trigger rule.
it should be sufficient to check the trigger rule.
and with that, it may be unnecessary to create this variable

hussein-awala committed:
- Fix a bug with mapped tasks
- update the method which find all upstream setup tasks
- update some code according to code review

hussein-awala committed:
Merge branch 'main' into fix/setup_task_deps

hussein-awala committed:
Fix a bug and failed tests

dstandish committed:
reduce diffs by making setup_upstream_tasks optional

dstandish committed:
if a task has no upstreams, there cannot be an upstream setup task, even indirect

dstandish commented on tests/models/test_taskinstance.py:
```diff
@@ -1270,20 +1338,142 @@ def test_depends_on_past(self, dag_maker):
+            ),
         ],
     )
     def test_check_task_dependencies(
```
i think that some of these tests, where you've made substantial changes, i think we should just make a new test. 

it's very important that we're confident that we're not breaking anything here, so rather than make such significant changes to the test, i think it's preferable to add a new one alongside.  i'll try to do this right now.

hussein-awala replied:
It was my first thought, but I tried to avoid duplicating the code. But yeah +1

dstandish committed:
add indirect upstream tests as separate test

dstandish committed:
Merge branch 'main' into fix/setup_task_deps

dstandish committed:
remove new tests from existing test

dstandish commented on tests/models/test_taskinstance.py:
```diff
@@ -1103,59 +1103,116 @@ def test_depends_on_past(self, dag_maker):
     #   successes, skipped, failed, upstream_failed, removed, done
     @pytest.mark.parametrize(
-        "trigger_rule, upstream_setups, upstream_states, flag_upstream_failed, expect_state, expect_passed",
+        "trigger_rule, direct_upstream_setups, indirect_upstream_setups, upstream_states,"
```
there are so many changes to tests it makes it a bit hard to review.

dstandish commented on airflow/ti_deps/deps/trigger_rule_dep.py:
```diff
@@ -51,6 +52,7 @@ class _UpstreamTIStates(NamedTuple):
     done: int
     success_setup: int
     skipped_setup: int
+    failed_setup: int
```
until now, everything in this class is *direct* upstreams. is failed_setup direct only or does it include indirect too? if it includes indirect too, it should probably be clarified through a more precise variable name.  but perhaps better would be to avoid mixing direct and indirect in the same class if it can be avoided.  perhaps we can just add the information through an optional argument in `calculate` or something.  this would also make the diff easier to deal with.

dstandish replied:
the other issue is, there are more states than just "failed" that we need to account for.  there is also upstream_failed, skipped, etc -- essentially anything other than success.

hussein-awala replied:
>  is failed_setup direct only or does it include indirect too?

yes it includes indirect too. The same for success_setup and skipped_setup

hussein-awala replied:
For me there is 4 different cases:
- running (all the unfinished states) -> we should wait
- success -> we can run the task
- failed (failed or upstream_failed) -> fail
- skipped with any fail -> we should skipp
So failed here includes failed and upstream_failed:
```python
        return _UpstreamTIStates(
            success=counter.get(TaskInstanceState.SUCCESS, 0),
            skipped=counter.get(TaskInstanceState.SKIPPED, 0),
            failed=counter.get(TaskInstanceState.FAILED, 0),
            upstream_failed=counter.get(TaskInstanceState.UPSTREAM_FAILED, 0),
            removed=counter.get(TaskInstanceState.REMOVED, 0),
            done=sum(counter.values()),
            success_setup=setup_counter.get(TaskInstanceState.SUCCESS, 0),
            skipped_setup=setup_counter.get(TaskInstanceState.SKIPPED, 0),
            failed_setup=setup_counter.get(TaskInstanceState.FAILED, 0)
            + setup_counter.get(TaskInstanceState.UPSTREAM_FAILED, 0),
        )
```

dstandish commented on airflow/ti_deps/deps/trigger_rule_dep.py:
```diff
@@ -94,33 +98,41 @@ def _get_dep_statuses(
 
         :param ti: Task instance to evaluate the trigger rule of.
         :param dep_context: The current dependency context.
+        setup_upstream_tasks: The setup tasks upstream of the current task.
```
should probably clarify that it includes indirect

dstandish commented on airflow/ti_deps/deps/trigger_rule_dep.py:
```diff
@@ -94,33 +98,41 @@ def _get_dep_statuses(
             return
-        if ti.task.trigger_rule == TR.ALWAYS:
+        if ti.task.trigger_rule == TR.ALWAYS and not setup_upstream_tasks:
+            # even with ALWAYS trigger rule, we still need to check setup tasks
```
i think always should probably keep the same behavior.  it says in the docs `No dependencies at all, run this task at any time`.  it used to be called "dummy" trigger rule.  so i think we just keep the absolute short circuit here.

dstandish replied:
it's really a non-sensical rule. with it, the deps are just for show. no one should ever really use this trigger rule.  perhaps it's there for testing 🤷

dstandish commented on airflow/ti_deps/deps/trigger_rule_dep.py:
```diff
@@ -94,33 +98,41 @@ def _get_dep_statuses(
             yield self._passing_status(reason="The task had a always trigger rule set.")
             return
-        yield from self._evaluate_trigger_rule(ti=ti, dep_context=dep_context, session=session)
+        yield from self._evaluate_trigger_rule(
```
given that there's no need to evaluate whether there are upstream setups for the purpose of TR.ALWAYS, i think we no longer need this new param `setup_upstream_tasks` in `_evaluate_trigger_rule`

dstandish committed:
remove setup_upstream_tasks from evaluate_trigger_rule and restore TR.ALWAYS behavior

dstandish committed:
remove comment

hussein-awala committed:
fix trigger_rule_dep test

hussein-awala committed:
Split upstream and setup upstream in two lists to fix a bug in trigger rule dep check

hussein-awala committed:
Make direct setup task respect trigger rule

hussein-awala committed:
Check setup tasks before the direct upstream tasks

dstandish commented on airflow/models/dag.py:
```diff
@@ -717,6 +718,16 @@ def validate_setup_teardown(self):
+                for down_task in task.downstream_list:
+                    if not down_task.is_teardown and down_task.trigger_rule not in [
+                        TriggerRule.ALL_SUCCESS,
+                        TriggerRule.ONE_SUCCESS,
```
```suggestion
```

dstandish commented on airflow/models/dag.py:
```diff
@@ -717,6 +718,16 @@ def validate_setup_teardown(self):
+                    ]:
+                        # this is required to ensure consistent clearing behavior when upstream
+                        raise ValueError(
+                            "Setup tasks must be followed with trigger rule ALL_SUCCESS or ONE_SUCCESS."
```
```suggestion
                            "Setup tasks must be followed with trigger rule ALL_SUCCESS."
```

dstandish committed:
revert some changes to tests for easier review

hussein-awala committed:
Apply suggestions from code review

Co-authored-by: Daniel Standish <15932138+dstandish@users.noreply.github.com>

hussein-awala committed:
fix unit tests and static checks

dstandish committed:
remove comment

hussein-awala committed:
calculate number of setup tasks which we should wait for from TIs instead of tasks list

dstandish committed:
small simplification

dstandish committed:
simplify by using existing structure and convert to set

hussein-awala committed:
Optimise the perf by getting the count of the relevant TI instead of getting all tasks

hussein-awala committed:
Check if the new method fixes the tests

uranusjr commented on airflow/models/dag.py:
```diff
@@ -717,6 +718,11 @@ def validate_setup_teardown(self):
+                for down_task in task.downstream_list:
+                    if not down_task.is_teardown and down_task.trigger_rule != TriggerRule.ALL_SUCCESS:
+                        # this is required to ensure consistent clearing behavior when upstream
+                        raise ValueError("Setup tasks must be followed with trigger rule ALL_SUCCESS.")
```
Since we’ve drilled this deep, can this show what the offending task is for clarity?

uranusjr commented on airflow/ti_deps/deps/trigger_rule_dep.py:
```diff
@@ -34,10 +36,13 @@
 
 
-class _UpstreamTIStates(NamedTuple):
+@dataclass
+class _UpstreamTIStates:
```
Is this just to have the default? Dataclass is pretty significantly slower and not really worthwhile here since the class is unpacked pretty much immediately. Better to stick to a named tuple.

uranusjr commented on airflow/ti_deps/deps/trigger_rule_dep.py:
```diff
@@ -121,6 +132,13 @@ def _evaluate_trigger_rule(
         from airflow.models.operator import needs_expansion
         from airflow.models.taskinstance import TaskInstance
 
+        if ti.task.is_teardown:
+            setup_upstream_tasks = [task for task in ti.task.upstream_list if task.is_setup]
+        else:
+            setup_upstream_tasks = list(ti.task.get_upstreams_only_setups())
```
```suggestion
        if ti.task.is_teardown:
            setup_upstream_tasks = (task for task in ti.task.upstream_list if task.is_setup)
        else:
            setup_upstream_tasks = ti.task.get_upstreams_only_setups()
```

No need to build a list here as far as I can tell

uranusjr commented on airflow/ti_deps/deps/trigger_rule_dep.py:
```diff
@@ -234,59 +296,72 @@ def _iter_upstream_conditions() -> Iterator[ColumnOperators]:
+            elif not task.is_teardown and upstream_setup and setup_done and skipped_setup > 0:
+                # when there are upstream setup tasks and at least one of them is skipped, then skip
+                new_state = TaskInstanceState.SKIPPED
+            elif not upstream_setup or setup_done:
+                # if there are no upstream setup tasks or all of them are done,
+                # and we haven't set a new state, then we can check the upstream tasks
```
I’d do

```python
elif upstream_setup and not setup_down:
    pass
```

and dedent the entire block below. (Not sure if I got the boolean negation right)
